python/ray/serve/_private/deployment_state.py

```diff
@@ -2041,23 +2037,37 @@ def _check_startup_replicas(
                 else:
                     self._replicas.add(original_state, replica)
 
+        if replicas_failed:
+            self.update_replica_startup_backoff_time()
+
+        return slow_replicas
+
+    def record_replica_startup_failure(self, error_msg: str):
+        """Record the error message for a replica startup failure.
+
+        Updates internal replica startup failure counter.
+        """
+
+        if self._replica_constructor_retry_counter >= 0:
+            # Increase startup failure counter if we're tracking it
+            self._replica_constructor_retry_counter += 1
+            self._replica_constructor_error_msg = error_msg
```

Comment:
I know this is just a refactor, but seems like there could be multiple replicas failing for different reasons? Perhaps we should change the default of `self._replica_constructor_error_msg` to empty string and append new error message when apply?

Reply:
Yeah, this is a good point. I'm on the fence about appending different error messages. Typically, the error messages originates [here](https://github.com/ray-project/ray/blob/8efd92074507ee60674dcaf94f5fb652c101e630/python/ray/serve/_private/deployment_state.py#L637-L652) and provides a full traceback, so users can debug. However, that means these errors aren't short, and appending multiple together could make it extra confusing.

We typically log the exceptions anyways ([example](https://github.com/ray-project/ray/blob/8efd92074507ee60674dcaf94f5fb652c101e630/python/ray/serve/_private/deployment_state.py#L638-L640)), so users will have access to them in the controller logs. I'd prefer to emit just the latest reason as a part of the status for readability's sake.

Reply:
Yea and that's the existing behavior anyways. We can keep as is until if user complains